File: week07/kjg/n-queen.js
function solution(n) {
    let answer = 0;
    const cols = Array(n).fill(0);

    function isValid(row, col) {
        for (let i = 0; i < row; i++) {
            if (cols[i] === col) return false;
            if (Math.abs(row - i) === Math.abs(col - cols[i])) return false;
        }
        return true;
    }

    function dfs(row) {
        if (row === n) {
            answer++;
            return;
        }

        for (let col = 0; col < n; col++) {
            if (isValid(row, col)) {
                cols[row] = col;
                dfs(row + 1);
            }
        }
    }

    dfs(0);
    return answer;
}

File: week07/ohy/N-Queen.java
class Solution {
    
    static int answer;
    static boolean[] colVisisted, leftCrossVisited, rightCrossVisited;
    
    private void dfs(int depth, int n) {
        if (depth == n) {
            answer++;
            return;
        }
        
        
        for (int col = 0; col < n; col++) {
            int left = depth - col + (n - 1);
            int right = depth + col;

            
            if(!colVisisted[col] && !leftCrossVisited[left] && !rightCrossVisited[right]) {
                colVisisted[col] = true;
                leftCrossVisited[left] = true;
                rightCrossVisited[right] = true;
                
                dfs(depth + 1, n);
                
                colVisisted[col] = false;
                leftCrossVisited[left] = false;
                rightCrossVisited[right] = false;
            }
        }
    }
    
    
    public int solution(int n) {
        answer = 0;
        colVisisted = new boolean[n];
        leftCrossVisited = new boolean[n * 2 - 1];
        rightCrossVisited = new boolean[n * 2 - 1];
        
        dfs(0, n);
        
        return answer;
    }
}
